In [2]:
import pandas as pd
import pickle

In [3]:
combined_df = pd.read_pickle('../../data/combined_analysis_data.pkl')
combined_df.head()

lda_0     lda_1     lda_2     lda_3     lda_4     lda_5     lda_6  \
index                                                                           
2490481 -0.469738 -0.450741 -0.460415 -0.514128 -0.468683 -0.465732 -0.445607   
2490482 -0.469738 -0.450741 -0.460415  0.751681 -0.468683 -0.465732  1.748294   
2490483 -0.346310 -0.323680 -0.339484 -0.415414 -0.356224 -0.328610 -0.282506   
2490484 -0.469738 -0.450741  6.469216 -0.514128 -0.468683 -0.465732 -0.445607   
2490485 -0.469738 -0.450741 -0.460415  4.390862 -0.468683 -0.465732 -0.445607   

            lda_7     lda_8     lda_9  ...  Flipping_the_Narrative  downvoted  \
index                                  ...                                      
2490481 -0.416617 -0.387825  1.437807  ...               -0.034424  -0.098938   
2490482 -0.416617 -0.387825 -0.455476  ...               -0.034424  -0.098938   
2490483 -0.227255 -0.186713 -0.293744  ...               -0.034424  -0.098938   
2490484 -0.416617 -0.387825 -0.455476  ...               -0.034424  -0.098938   
2490485 -0.416617 -0.387825 -0.455476  ...               -0.034424  -0.098938   

         moderator  downvoted_user  deleted_user  log_nword  user_flair  \
index                                                                     
2490481  -0.011314       -0.046315     -0.165895   0.679459    1.082047   
2490482  -0.011314       -0.046315     -0.165895   0.056112    1.082047   
2490483  -0.011314       -0.046315     -0.165895  -0.639216    1.082047   
2490484  -0.011314       -0.046315     -0.165895  -0.179831    1.082047   
2490485  -0.011314       -0.046315     -0.165895   0.348764    1.082047   

         pt_user_count  pt_user_mean  pt_score  
index                                           
2490481      -0.214914     -0.420876 -1.207604  
2490482       0.247121      1.785519  0.171363  
2490483       0.298072      0.829723 -0.112168  
2490484       0.984762      0.840029  0.885040  
2490485       0.450655      1.779131  0.790891  

[5 rows x 39 columns]

In [4]:
from sklearn.cluster import KMeans

plot_kmk = []
mod_kmk = []
for k in range(1,30):
    km_k = KMeans(n_clusters = k)
    km_k.fit(combined_df)
    plot_kmk.append(km_k.inertia_)
    mod_kmk.append(km_k)

/home/nick/miniconda3/envs/CAPP/lib/python3.9/site-packages/sklearn/cluster/_kmeans.py:94: RuntimeWarning: divide by zero encountered in log
  n_local_trials = 2 + int(np.log(n_clusters))


OverflowError: cannot convert float infinity to integer

In [1]:
import matplotlib.pyplot as plt

plt.plot(plot_kmk)
plt.title('Elbow Plot')
plt.xlabel('Number of Clusters')
plt.ylabel('Sum Squared Distance')

NameError: name 'plot_kmk' is not defined

Even using 20 clusters, we aren't getting a substantial reduction in SSD - certainly not one that makes any real improvement on the LDA inputs we got from the text in the first place.

In [4]:
#lets try another clustering approach - agglomerative
from sklearn.cluster import AgglomerativeClustering
agg = AgglomerativeClustering()
agg.fit(combined_df)

MemoryError: Unable to allocate 8.27 TiB for an array with shape (1137117203596,) and data type float64

Agglomerative clustering needs to allocate an array of size 8.27 Terabytes... this isn't going to happen. 

In [ ]:
#trying a density approach - DBScan
from sklearn.cluster import DBSCAN
from sklearn.metrics import silhouette_score

eps_opts = [0.01, 0.05, 0.1, 0.2]
minsamples_opts = [100, 1000, 10000, 100000]

plot_db = []
mod_kmk = []
for eps in eps_opts:
    for minsamples in minsamples_opts:
        db_tst = DBSCAN(eps=eps, min_samples=minsamples)
        db_tst.fit_predict(combined_df)
        plot_kmk.append(km_k.inertia_)
        mod_kmk.append(km_k)

In [7]:
#trying a density approach - DBScan
from sklearn.cluster import DBSCAN
from sklearn.metrics import silhouette_score
dbscan = DBSCAN()
db_tst = DBSCAN(eps = .1, min_samples = 1000)
db_tst.fit_predict(combined_df)

KeyboardInterrupt: 